### PR TITLE
Search package: search plan update on heartbeat and new method `ever_supported_search`

### PR DESCRIPTION
**Pull request by kangzj**
#### Changes proposed in this Pull Request:
The PR proposes changes to add `ever_supported_search` for the determination of Search submenu item visibility, i.e. if the site ever had/have a plan supporting Search then the menu is shown. The PR also added plan data updating on Jetpack heartbeat.

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Run `jetpack install --all`
* `(cd projects/packages/search && composer phpunit)`
* Ensure all tests pass

### CODE REVIEW

File: projects/packages/search/src/class-plan.php [hunk 1]
```diff
@@ -16,7 +16,27 @@
...
+	 * @var boolean
+	 */
+	protected static $hooks_inited = false;
+
+	/**
+	 * Init hooks for updating plan info
+	 */
+	public function init_hooks() {
+		// Update plan info from WPCOM on Jetpack heartbeat.
+		// TODO: implement heartbeart for search.
```
Comment (jsnmoon) [inline projects/packages/search/src/class-plan.php:34]: Is this comment correct? I thought this PR implemented the heartbeat?

If this PR doesn't implement the heartbeat, what does it do exactly?

File: projects/packages/search/src/class-plan.php [hunk 2]
```diff
@@ -16,7 +16,27 @@
...
+		if ( ! static::$hooks_inited ) {
+			add_action( 'jetpack_heartbeat', array( $this, 'get_plan_info_from_wpcom' ) );
+			static::$hooks_inited = true;
+		}
```
Comment (jsnmoon) [inline projects/packages/search/src/class-plan.php:38]:
```suggestion
		if ( ! static::$heartbeat_initialized ) {
			add_action( 'jetpack_heartbeat', array( $this, 'get_plan_info_from_wpcom' ) );
			static::$heartbeat_initialized = true;
		}
```

I recommend using a more descriptive variable name like above here.

**Review by jsnmoon — COMMENTED** [review]
I agree with using the heartbeat approach for determining plan eligibility 👍 

> The PR proposes changes to add ever_supported_search for the determination of Search submenu item visibility, i.e. if the site ever had/have a plan supporting Search then the menu is shown.

I think this is a good start. In the future, we may want to consider showing the submenu even if they've never had a search plan in the past, especially for customers who have just installed the standalone Search plugin.

File: projects/packages/search/src/class-plan.php (hunk 2 above)
Comment (kangzj) [reply projects/packages/search/src/class-plan.php]: Changed to `$update_plan_hook_initialized` in eaf94573

File: projects/packages/search/src/class-plan.php (hunk 1 above)
Comment (jsnmoon) [reply projects/packages/search/src/class-plan.php:34]: Ah, got it. Let me rephrase what you've said to see if I understood you correctly.

This PR doesn't implement heartbeat for Search because we're not emitting any data back to WPCOM servers. However, it does add periodic polling of WPCOM plan info.

Is this correct?

Comment (kangzj) [reply projects/packages/search/src/class-plan.php:34]: >This PR doesn't implement heartbeat for Search because we're not emitting any data back to WPCOM servers. However, it does add periodic polling of WPCOM plan info.

hmmm, no. In other words, Jetpack implemented its heartbeat ( it applies 'jetpack_heartbeat' action and also provide an endpoint for WPCOM to call ). The PR hooks 'jetpack_heartbeat' to update search plan data. When we start our standalone Search plugin, there will not be such heartbeat, which is the thing we'd need to implement.

Does that make sense?

Comment (jsnmoon) [reply projects/packages/search/src/class-plan.php:34]: Ah, got it! Thanks for the clarification, that makes much more sense.

**Review by jsnmoon — DISMISSED** [review]

**Comment by jsnmoon** [comment]
Needs a master merge or a rebase.

**Review by jsnmoon — APPROVED** [review]

**Comment by kangzj** [comment]
Hi @jsnmoon Thanks for your speedy review 👍
